backend/python-functions/data_ingest.py

```diff
@@ -0,0 +1,145 @@
+        "canonicalName": "Denver Reception & Diagnostic Center",
+        "facilityType": "State Prisons",
+        "state": "Colorado",
+        "covidCases": {
+          "2020-06-01": {
+            "popDeaths": 0,
+            "popTested": 253,
+            "popTestedPositive": 2
+          },
+          "2020-06-02": {
+            "popDeaths": 0,
+            "popTested": 253,
+            "popTestedPositive": 2
+          },
+          "2020-06-03": {
+            "popDeaths": 0,
+            "popTested": 253,
+            "popTestedPositive": 2
+          },
+          "2020-06-04": {
+            "popDeaths": 0,
+            "popTested": 253,
+            "popTestedPositive": 2
+          }
+        }
+      },
+      ...
+    }
+    """
+    facilities = {}    
+    with open(file_location) as csv_file:
+        reader = csv.DictReader(csv_file, delimiter=',')
+
+        for row in reader:
+            state = row["State"].strip()
+            canonical_facility_name = row["Canonical Facility Name"].strip()
+            facility_type = row['Facility Type'].strip()
+            key = f'{state}::{canonical_facility_name}'
+            
+            if(key in facilities):
+                # If the facility already exists, append the case counts for a given day.
+                facilities[key]['covidCases'][row['Date']] = build_covid_case_counts(row)
+            else:
+                # If the facility does not already exist, save its metadata along with
+                # the case counts for a given day.
+                facilities[key] = {
+                    'canonicalName': canonical_facility_name,
+                    'facilityType': facility_type,
+                    'state': state,
+                    'covidCases': {
+                        f'{row["Date"]}': build_covid_case_counts(row)
+                    }
+                }
+
+    return facilities
+
+def persist(facilities):
+    db = firestore.Client()
+
+    facilities_collection = db.collection(REFERENCE_FACILITIES_COLLECTION_ID)
+                
+    for key, facility in facilities.items():
+        facilitiesQueryResult = facilities_collection \
+            .where('state', '==', f'{facility["state"]}') \
+            .where('canonicalName', '==', f'{facility["canonicalName"]}') \
+            .stream()
+
+        facilityDocuments = list(facilitiesQueryResult)       
+
+        if len(facilityDocuments) == 0 or len(facilityDocuments) == 1:
+            # Accesses an existing reference facility document or creates a new one if
+            # one does not exist.
+            facility_id = facilityDocuments[0].id if len(facilityDocuments) == 1 else None
+            facility_ref = facilities_collection.document(facility_id)
+            
+            # Remove the Covid case data so that it can be stored separately in its own
+            # sub-collection.
+            covid_cases = facility.pop('covidCases')
+            
+            batch = db.batch()
```

Comment:
For each facility, we write its metadata and its daily case counts in a single batch request.  By performing these database writes in a batch, we limit the number of requests we need to send to match the number of facilities within a CSV file and **_not_** the number of daily case rows.  For example, at the moment this means we'll only be performing ~1350 requests for the entire ~50k line CSV file.  Additionally, this means that even as the file grows each day to add new case counts, we won't be generating more requests.  We'll only add more requests as we add more facilities, which should be must less frequent.

Reply:
The batch also speeds things up greatly.  Before performing these operations within a batch, the function could not finish executing within the allotted timeout (9 minutes, the max).  With the batch in place, this function completes in about 2.5 minutes.

Reply:
Fantastic!